### PR TITLE
Iterators: Implement ops to convert values to and from streams.

### PR DESCRIPTION
**Pull request by ingomueller-net**
This PR ~~depends on and therefore~~ includes #616 and #617.

### CODE REVIEW

File: experimental/iterators/include/iterators/Dialect/Iterators/IR/IteratorsOps.td [hunk 1]
```diff
@@ -303,6 +303,52 @@ def Iterators_SinkOp : Iterators_Base_Op<"sink"> {
...
+    Consumes the first element of the given input stream and returns it. The
+    remaining elements from the input stream are not consumed.
+
+    The purpose of this op is to pass from "stream land" to "value land," i.e.,
+    to enable non-iterator ops to consume the results of a tree of iterator ops.
+
+    Example:
+    ```mlir
+    %stream = ...
+    %value = iterators.stream_to_value %stream : <i32>
```
Comment (nicolasvasilache) [inline experimental/iterators/include/iterators/Dialect/Iterators/IR/IteratorsOps.td:321]: type looks weird here

File: experimental/iterators/lib/Conversion/IteratorsToLLVM/IteratorsToLLVM.cpp [hunk 2]
```diff
@@ -1158,6 +1158,92 @@ static Value buildStateCreation(TabularViewToStreamOp op,
...
+static llvm::SmallVector<Value, 4>
+buildNextBody(ValueToStreamOp op, OpBuilder &builder, Value initialState,
+              ArrayRef<IteratorInfo> upstreamInfos, Type elementType) {
+  Location loc = op.getLoc();
+  ImplicitLocOpBuilder b(loc, builder);
+  Type i1 = b.getI1Type();
+
+  // Check if the iterator has returned an element already (since it should
+  // return one only in the first call to next).
+  Value hasReturned =
```
Comment (nicolasvasilache) [inline experimental/iterators/lib/Conversion/IteratorsToLLVM/IteratorsToLLVM.cpp:1204]: this looks redundant with other things, can this be extracted to a common place and take a lambda somewhere ?

File: experimental/iterators/lib/Conversion/IteratorsToLLVM/IteratorsToLLVM.cpp [hunk 3]
```diff
@@ -1621,6 +1769,12 @@ static void convertIteratorOps(ModuleOp module, TypeConverter &typeConverter) {
                  "Expected iterator op to be converted to one value.");
           mapping.map(op->getResult(0), converted[0]);
         })
+        .Case<StreamToValueOp>([&](auto op) {
```
Comment (nicolasvasilache) [inline experimental/iterators/lib/Conversion/IteratorsToLLVM/IteratorsToLLVM.cpp:1772]: Can we add some comments in this case and the previous one on the types that are expected before/after conversion?

File: experimental/iterators/test/Conversion/IteratorsToLLVM/stream-to-value.mlir [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+// CHECK-LABEL: func.func private @iterators.value_to_stream.open.{{[0-9]+}}(%{{.*}}: !iterators.state<i1, i32>) -> !iterators.state<i1, i32> {
+
+func.func @main() {
+// CHECK-LABEL: func.func @main() {
+  %value = arith.constant 42 : i32
+// CHECK-NEXT:    %[[V0:.*]] = arith.constant 42 : i32
+  %stream = iterators.value_to_stream %value : <i32>
+// CHECK-NEXT:    %[[V1:.*]] = iterators.undefstate : !iterators.state<i1, i32>
+  %result:2 = iterators.stream_to_value %stream : <i32>
+// CHECK-NEXT:    %[[V2:.*]] = iterators.insertvalue %[[V0]] into %[[V1]][1] : !iterators.state<i1, i32>
```
Comment (nicolasvasilache) [inline experimental/iterators/test/Conversion/IteratorsToLLVM/stream-to-value.mlir:17]: this corresponds to the previous line ?
`= iterators.value_to_stream %value : <i32>`

**Review by nicolasvasilache — APPROVED** [review]

File: experimental/iterators/include/iterators/Dialect/Iterators/IR/IteratorsOps.td (hunk 1 above)
Comment (ingomueller-net) [reply experimental/iterators/include/iterators/Dialect/Iterators/IR/IteratorsOps.td]: Solved in latest commit.

File: experimental/iterators/test/Conversion/IteratorsToLLVM/stream-to-value.mlir (hunk 4 above)
Comment (ingomueller-net) [reply experimental/iterators/test/Conversion/IteratorsToLLVM/stream-to-value.mlir:16]: Fixed in latest commit.

File: experimental/iterators/lib/Conversion/IteratorsToLLVM/IteratorsToLLVM.cpp (hunk 3 above)
Comment (ingomueller-net) [reply experimental/iterators/lib/Conversion/IteratorsToLLVM/IteratorsToLLVM.cpp:1773]: Solved in latest commit.

**Pull request merged**